programs/marginfi/src/state/emode.rs

```diff
@@ -1,35 +1,125 @@
 use anchor_lang::err;
 use fixed::types::I80F48;
-use marginfi_type_crate::types::{EmodeSettings, EMODE_ON};
+use fixed_macro::types::I80F48;
+use marginfi_type_crate::types::{BankConfig, EmodeSettings, EMODE_ON};
 
-use crate::{check, errors::MarginfiError, prelude::MarginfiResult};
+use crate::{
+    check,
+    errors::MarginfiError,
+    math_error,
+    prelude::MarginfiResult,
+    state::{bank_config::BankConfigImpl, marginfi_account::RequirementType},
+};
+use marginfi_type_crate::types::u32_to_basis;
+
+/// Default Maximum allowed theoretical leverage for emode configurations (initial).
+/// L = 1 / (1 - CW/LW) where CW is collateral weight and LW is liability weight.
+/// A value of 15 means positions can theoretically leverage up to 15x through recursive borrowing.
+pub const DEFAULT_INIT_MAX_EMODE_LEVERAGE: I80F48 = I80F48!(15);
+
+/// Default Maximum allowed theoretical leverage for emode configurations (maintenance).
+/// L = 1 / (1 - CW/LW) where CW is collateral weight and LW is liability weight.
+/// A value of 20 means positions can theoretically leverage up to 20x through recursive borrowing.
+pub const DEFAULT_MAINT_MAX_EMODE_LEVERAGE: I80F48 = I80F48!(20);
 
 pub trait EmodeSettingsImpl {
-    fn validate_entries(&self) -> MarginfiResult;
+    fn validate_entries_with_liability_weights(
+        &self,
+        bank_config: &BankConfig,
+        emode_max_init_leverage: u32,
+        emode_max_maint_leverage: u32,
+    ) -> MarginfiResult;
     fn check_dupes(&self) -> MarginfiResult;
     fn is_enabled(&self) -> bool;
     fn update_emode_enabled(&mut self);
 }
 
+/// Calculate theoretical maximum leverage given collateral and liability weights.
+/// Formula: L = 1 / (1 - CW/LW)
+///
+/// # Arguments
+/// * `collateral_weight` - The collateral weight (CW)
+/// * `liability_weight` - The liability weight (LW)
+///
+/// # Returns
+/// * Ok(leverage) if calculation is valid (CW < LW)
+/// * Err if leverage would be infinite or negative (CW >= LW)
+#[inline]
+pub fn calculate_max_leverage(
+    collateral_weight: I80F48,
+    liability_weight: I80F48,
+) -> MarginfiResult<I80F48> {
+    // Ensure liability weight is positive
+    check!(
+        liability_weight > I80F48::ZERO,
+        MarginfiError::BadEmodeConfig
+    );
+
+    // Ensure collateral weight < liability weight (strictly less than)
+    check!(
+        collateral_weight < liability_weight,
+        MarginfiError::BadEmodeConfig
+    );
+
+    //  ratio =  CW/LW
+    let ratio: I80F48 = collateral_weight
+        .checked_div(liability_weight)
```

Comment:
restored the checked div, I forgot that this is technically mock floating point math and (big_number)/(very_small_fractional_number) can overflow so we need checked_div even if the denominator is known not zero.